climada_petals/hazard/tc_surge_bathtub.py

```diff
@@ -119,17 +122,16 @@ def from_tc_winds(wind_haz, topo_path, inland_decay_rate=0.2, add_sea_level_rise
         coastal_centroids_h -= add_sea_level_rise
 
         # Efficient way to subtract from selected columns of sparse csr matrix
-        nz_coastal_cents = inten_surge[:,coastal_idx].nonzero()
-        nz_coastal_cents_inten = (nz_coastal_cents[0], coastal_idx[nz_coastal_cents[1]])
-        inten_surge[nz_coastal_cents_inten] -= coastal_centroids_h[nz_coastal_cents[1]]
+        inten_surge.data -= coastal_centroids_h[cent_to_coastal_idx[inten_surge.indices]]
 
         # Discard negative (invalid/unphysical) surge height values
         inten_surge.data = np.fmax(inten_surge.data, 0)
         inten_surge.eliminate_zeros()
 
         # Get fraction of (large) centroid cells on land according to the given (high-res) DEM
-        fract_surge = sp.csr_matrix(_fraction_on_land(centroids, topo_path))
-        fract_surge = sp.csr_matrix(np.ones((inten_surge.shape[0], 1))) * fract_surge
+        # only store the result in locations with non-zero surge height
+        fract_surge = inten_surge.copy()
+        fract_surge.data[:] = _fraction_on_land(centroids, topo_path)[fract_surge.indices]
```

Comment:
This not only reduces the amount of memory required for computation, but also reduces the effective amount of data stored in the result object. It will only store a non-zero `fraction` value for events and centroids where a non-zero `intensity` value is stored. Since the `fraction` is never used for other centroids, this was a huge waste of storage in the previous version, and would often produce out-of-memory errors.